File: backend/app/models/reminder.py
import uuid
from pydantic import ConfigDict
from sqlalchemy import (
    Column,
    String,
    Text,
    Boolean,
    Integer,
    DateTime,
    ForeignKey,
    JSON,
)
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.sql import func
from app.db import Base


class Reminder(Base):
    __tablename__ = "reminders"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)

    user_id = Column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )

    title = Column(String(255), nullable=False)
    description = Column(Text)

    type = Column(String(50), nullable=False)
    # birthday | anniversary | task | bill | health | trip | custom

    start_at = Column(DateTime(timezone=True), nullable=False)
    end_at = Column(DateTime(timezone=True))

    is_all_day = Column(Boolean, default=False)
    timezone = Column(String(50), nullable=False)

    recurrence_rule = Column(Text)
    # RFC5545 RRULE

    notify_offsets = Column(JSON, default=list)
    # seconds before event

    priority = Column(Integer, default=0)
    status = Column(String(20), default="active")
    # active | completed | cancelled

    completed_at = Column(DateTime(timezone=True))

    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(
        DateTime(timezone=True),
        server_default=func.now(),
        onupdate=func.now(),
    )

    model_config = ConfigDict(from_attributes=True)

File: mobile/app/lib/api/models/reminder.dart
import '../models/reminder_type.dart';

class Reminder {
  final String? id;
  final String title;
  final String? description;
  final ReminderType type;
  final DateTime startAt;
  final DateTime? endAt;
  final bool isAllDay;
  final String timezone;
  final String? recurrenceRule;
  final List<int> notifyOffsets;
  final int priority;
  final String status;
  final DateTime? completedAt;

  Reminder({
    this.id,
    required this.title,
    this.description,
    required this.type,
    required this.startAt,
    this.endAt,
    this.isAllDay = false,
    required this.timezone,
    this.recurrenceRule,
    this.notifyOffsets = const [],
    this.priority = 0,
    this.status = 'active',
    this.completedAt,
  });

  factory Reminder.fromJson(Map<String, dynamic> json) {
    return Reminder(
      id: json['id'] as String?,
      title: json['title'] as String,
      description: json['description'] as String?,
      type: ReminderTypeExtension.fromString(json['type'] as String),
      startAt: DateTime.parse(json['start_at'] as String),
      endAt: json['end_at'] != null
          ? DateTime.parse(json['end_at'] as String)
          : null,
      isAllDay: json['is_all_day'] as bool? ?? false,
      timezone: json['timezone'] as String,
      recurrenceRule: json['recurrence_rule'] as String?,
      notifyOffsets: (json['notify_offsets'] as List<dynamic>?)
              ?.map((e) => e as int)
              .toList() ??
          [],
      priority: json['priority'] as int? ?? 0,
      status: json['status'] as String? ?? 'active',
      completedAt: json['completed_at'] != null
          ? DateTime.parse(json['completed_at'] as String)
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      if (id != null) 'id': id,
      'title': title,
      if (description != null) 'description': description,
      'type': type.name,
      'start_at': startAt.toUtc().toIso8601String(),
      if (endAt != null) 'end_at': endAt!.toUtc().toIso8601String(),
      'is_all_day': isAllDay,
      'timezone': timezone,
      if (recurrenceRule != null) 'recurrence_rule': recurrenceRule,
      'notify_offsets': notifyOffsets,
      'priority': priority,
      'status': status,
      if (completedAt != null)
        'completed_at': completedAt!.toUtc().toIso8601String(),
    };
  }
}
